File: 3_matrix_mult_logical_loops.c
#include <stdio.h>

#define ROWS_A 3
#define COLS_A 2
#define ROWS_B 2
#define COLS_B 3

// Print matrix
void print_matrix(int *matrix, int rows, int cols) {
    int i = 0;
    while (i < rows) {
        int j = 0;
        while (j < cols) {
            printf("%d ", *(matrix + i * cols + j));
            j++;
        }
        printf("\n");
        i++;
    }
}

// Matrix multiplication using logical loops (while)
void multiply_matrices_logical(int A[ROWS_A][COLS_A], 
                               int B[ROWS_B][COLS_B],
                               int result[ROWS_A][COLS_B]) {
    int i = 0;
    while (i < ROWS_A) {
        int j = 0;
        while (j < COLS_B) {
            result[i][j] = 0;
            int k = 0;
            while (k < COLS_A) {
                result[i][j] += A[i][k] * B[k][j];
                k++;
            }
            j++;
        }
        i++;
    }
}

int main() {
    int A[ROWS_A][COLS_A] = {
        {1, 2},
        {3, 4},
        {5, 6}
    };

    int B[ROWS_B][COLS_B] = {
        {7, 8, 9},
        {10, 11, 12}
    };

    int result[ROWS_A][COLS_B];

    printf("=== Matrix Multiplication with Logical Loops (while) ===\n");
    printf("Matrix A (%d x %d):\n", ROWS_A, COLS_A);
    print_matrix((int *)A, ROWS_A, COLS_A);

    printf("\nMatrix B (%d x %d):\n", ROWS_B, COLS_B);
    print_matrix((int *)B, ROWS_B, COLS_B);

    multiply_matrices_logical(A, B, result);

    printf("\nResult (A * B) (%d x %d):\n", ROWS_A, COLS_B);
    print_matrix((int *)result, ROWS_A, COLS_B);

    return 0;
}

File: 3_matrix_mult_logical_loops.py
def print_matrix(matrix):
    """Print a matrix"""
    i = 0
    while i < len(matrix):
        j = 0
        while j < len(matrix[i]):
            print(matrix[i][j], end=" ")
            j += 1
        print()
        i += 1

def multiply_matrices_logical(A, B):
    """Matrix multiplication using logical loops (while)"""
    rows_A = len(A)
    cols_A = len(A[0])
    rows_B = len(B)
    cols_B = len(B[0])

    # Create result matrix with zeros
    result = [[0 for _ in range(cols_B)] for _ in range(rows_A)]

    # Multiply using while loops
    i = 0
    while i < rows_A:
        j = 0
        while j < cols_B:
            k = 0
            while k < cols_A:
                result[i][j] += A[i][k] * B[k][j]
                k += 1
            j += 1
        i += 1

    return result

# Test the logical loop implementation
A = [
    [1, 2],
    [3, 4],
    [5, 6]
]

B = [
    [7, 8, 9],
    [10, 11, 12]
]

print("=== Matrix Multiplication with Logical Loops (while) ===")
print("Matrix A (3 x 2):")
print_matrix(A)

print("\nMatrix B (2 x 3):")
print_matrix(B)

result = multiply_matrices_logical(A, B)

print("\nResult (A * B) (3 x 3):")
print_matrix(result)
